src/imlib.c

```diff
@@ -289,6 +297,124 @@ int feh_load_image(Imlib_Image * im, feh_file * file)
 	return(1);
 }
 
+static int feh_file_is_raw(char *filename)
+{
+	childpid = fork();
+	if (childpid == -1) {
+		perror("fork");
+		return 0;
+	}
+
+	if (childpid == 0) {
+		if (opt.quiet) {
+			int devnull = open("/dev/null", O_WRONLY);
+			dup2(devnull, 1);
+			dup2(devnull, 2);
+		}
+		execl("/bin/dcraw", "dcraw", "-i", filename, NULL);
+		_exit(1);
+	} else {
+		int status;
+		do {
+			waitpid(childpid, &status, WUNTRACED);
+			if (WIFEXITED(status)) {
+				return !WEXITSTATUS(status);
+			}
+		} while (!WIFEXITED(status) && !WIFSIGNALED(status));
+	}
+
+	return 0;
+}
+
+static char *feh_dcraw_load_image(char *filename)
+{
+	char *basename;
+	char *tmpname;
+	char *sfn;
+	int fd = -1;
+
+	basename = strrchr(filename, '/');
+
+	if (basename == NULL)
+		basename = filename;
+	else
+		basename++;
+
+	tmpname = feh_unique_filename("/tmp/", basename);
+
+	if (strlen(tmpname) > (NAME_MAX-6))
+		tmpname[NAME_MAX-7] = '\0';
+
+	sfn = estrjoin("_", tmpname, "XXXXXX", NULL);
+	free(tmpname);
+
+	fd = mkstemp(sfn);
+
+	if (fd == -1) {
+		free(sfn);
+		return NULL;
+	}
+
+	int filedes[2];
+	if (pipe(filedes) == -1) {
+		unlink(sfn);
+		free(sfn);
+		return NULL;
+	}
+
+	childpid = fork();
+	if (childpid == -1) {
+		weprintf("%s: Can't load with dcraw. Fork failed:", filename);
+		unlink(sfn);
+		free(sfn);
+		return NULL;
+	} else if (childpid == 0) {
+
+		while ((dup2(filedes[1], STDOUT_FILENO) == -1) && (errno == EINTR)) {}
+		close(filedes[1]);
+		close(filedes[0]);
+
+		alarm(opt.dcraw_timeout);
+		execlp("dcraw", "dcraw", "-c", "-e", filename, NULL);
```

Comment:
This instructs `dcraw` to output to standard output - the pipe could be replaced by just `dup2`-ing the open `fd` onto the standard output then, getting rid of the `pipe` call entirely and the buffer copying.

Reply:
Great proposal! Today I learned something.